### PR TITLE
refactor: define typescript types for OfferCancel Transaction Model

### PR DESCRIPTION
**Pull request by mukulljangid**
## High Level Overview of Change
Defines typescript types for OfferCancel transaction.

### Context of Change
OfferCancel is an extension of CommonFields, and is exported in a type union with all other transaction types defined as Transaction.

We also will provide an optional verifyOfferCancel function for runtime form checking of OfferCancel transactions, should the users want to verify transaction form at runtime.

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

## Test Plan
Adds tests for `verifyOfferCancel`, which provides runtime form verification for `OfferCancel` transactions

### CODE REVIEW

**Comment by mvadari** [comment]
This PR title should have an active verb as the first word

**Comment by intelliot** [comment]
While we're on the topic: for consistency, I recommend using the [imperative mood](https://chris.beams.io/posts/git-commit/). So "Define" instead of "Defines".

**Comment by intelliot** [comment]
An "upgrade" 😀   to the page I linked above: I like [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) even more. I'm not pedantic about it, but it's nice to see.

**Review by natenichols — APPROVED** [review]
LGTM

**Comment by natenichols** [comment]
You'll need to address the conflict, but that should be pretty easy.

File: test/models/offerCancel.ts
Comment (JST5000) [inline test/models/offerCancel.ts:23]: Why do we cast the offer as any? (Curious)

File: src/models/transactions/offerCancel.ts
Comment (JST5000) [inline src/models/transactions/offerCancel.ts:19]: Should having an undefined offerSequence be an error? (This doc says that it's not an error if the specified offer doesn't exist, but it seems like it might require at least a proposed offer to cancel: https://xrpl.org/offercancel.html)

**Review by JST5000 — COMMENTED** [review]
Couple of curiosities before I approve (Not sure what the expected behavior should be)

File: src/models/transactions/offerCancel.ts
Comment (khancode) [reply src/models/transactions/offerCancel.ts:19]: Isn't this type verification unnecessary when we have a defined TypeScript type for it? Wouldn't TS catch this error at compile time?

Comment (intelliot) [reply src/models/transactions/offerCancel.ts:19]: The library can also be used by JavaScript apps, which won't get type safety. These verifications are for them.

Even so, I don't think they are necessary or even particularly helpful. I would be in favor of getting rid of these checks, but it's not a big deal. I'll leave it to @mukulljangid @mvadari @natenichols to decide :)

File: src/models/transactions/offerCancel.ts
Comment (khancode) [inline src/models/transactions/offerCancel.ts:13]: I think it makes more sense to have return type be `boolean` instead of `void`. We can `return true` at the end of this function.

Comment (khancode) [inline src/models/transactions/offerCancel.ts:13]: I think it makes more sense to have return type be `boolean` instead of `void`. We can `return true` at the end of this function.

File: test/models/offerCancel.ts
Comment (mvadari) [reply test/models/offerCancel.ts:23]: Otherwise, TypeScript would complain that it's not the right format. But we want it to check at runtime instead of at compile time, so we cast it as `any`, which basically tells TS to ignore this.

File: src/models/transactions/offerCancel.ts
Comment (khancode) [inline src/models/transactions/offerCancel.ts:10]: Can you also specify this is an optional function to call for users?

**Review by khancode — CHANGES_REQUESTED** [review]

File: src/models/transactions/offerCancel.ts
Comment (mvadari) [reply src/models/transactions/offerCancel.ts:19]: We've talked it over with enough folks that we think that the runtime checks are useful.

File: src/models/transactions/offerCancel.ts
Comment (mvadari) [reply src/models/transactions/offerCancel.ts:13]: I don't think it makes sense to `return true` if there's no path where you could `return false`.

File: src/models/transactions/index.ts
Comment (mvadari) [inline src/models/transactions/index.ts:3]: merge issue here

File: src/models/transactions/transaction.ts
Comment (mvadari) [inline src/models/transactions/transaction.ts:10]: ditto

File: test/models/offerCancel.ts
Comment (mvadari) [inline test/models/offerCancel.ts:40]: this should be in a separate test

**Review by mvadari — CHANGES_REQUESTED** [review]

File: src/models/transactions/offerCancel.ts
Comment (khancode) [reply src/models/transactions/offerCancel.ts:13]: Hmm that's true (no pun). I'm used to expecting verify functions returning a `boolean` as a success indication. I'm assuming users will wrap `verify` functions in a try catch and handle whatever error is thrown.

File: src/models/transactions/transaction.ts
Comment (mukulljangid) [reply src/models/transactions/transaction.ts]: Done!

File: src/models/transactions/index.ts
Comment (mukulljangid) [reply src/models/transactions/index.ts]: Fixed!

File: src/models/transactions/offerCancel.ts
Comment (JST5000) [inline src/models/transactions/offerCancel.ts:19]: I think if OfferSequence is undefined this should actually throw an error (I don't think the field is optional - https://xrpl.org/offercancel.html)

File: src/models/transactions/offerCancel.ts
Comment (khancode) [inline src/models/transactions/offerCancel.ts:2]: nit: remove semicolons

File: src/models/transactions/offerCancel.ts
Comment (khancode) [inline src/models/transactions/offerCancel.ts:6]: ditto

File: src/models/transactions/transaction.ts
Comment (khancode) [inline src/models/transactions/transaction.ts:6]: remove semicolons to keep it consistent

**Review by khancode — APPROVED** [review]
LGTM. Left some nit comments on removing semicolons.

File: src/models/transactions/offerCancel.ts
Comment (mukulljangid) [reply src/models/transactions/offerCancel.ts]: Done!

**Review by JST5000 — APPROVED** [review]

File: test/models/offerCancel.ts
Comment (khancode) [inline test/models/offerCancel.ts:27]:
```suggestion
describe('OfferCancel Transaction Verification', function () {
   let offer
   
   beforeEach(() => {
       offer = {
            Account: "rnKiczmiQkZFiDES8THYyLA2pQohC5C6EF",
            Fee: "10",
            LastLedgerSequence: 65477334,
            OfferSequence: 60797528,
            Sequence: 60797535,
            SigningPubKey: "0361BFD43D1EEA54B77CC152887312949EBF052997FBFFCDAF6F2653164B55B21...",
            TransactionType: "OfferCancel",
            TxnSignature: "30450221008C43BDCFC68B4793857CA47DF454C07E5B45D3F80E8E6785CAB9292...",
            date: "2021-08-06T21:04:11Z"
        } as any
   })

    it (`verifies valid OfferCancel`, () => {        
        assert.doesNotThrow(() => verifyOfferCancel(offer))        
    })
```

File: test/models/offerCancel.ts
Comment (khancode) [inline test/models/offerCancel.ts:80]:
```suggestion
        delete offer.OfferSequence
```

File: test/models/offerCancel.ts
Comment (khancode) [inline test/models/offerCancel.ts:60]:
```suggestion
        offer.OfferSequence = '99'
```

File: test/models/offerCancel.ts
Comment (khancode) [inline test/models/offerCancel.ts:44]:
```suggestion
        offer.Flags = 2147483648
        assert.doesNotThrow(() => verifyOfferCancel(offer))
```

**Review by khancode — CHANGES_REQUESTED** [review]
Left some suggested changes to clean up tests by using `beforeEach()`

File: test/models/offerCancel.ts
Comment (khancode) [inline test/models/offerCancel.ts:32]: nit: delete empty line

File: test/models/offerCancel.ts
Comment (khancode) [inline test/models/offerCancel.ts:40]: ditto

File: test/models/offerCancel.ts
Comment (khancode) [inline test/models/offerCancel.ts:50]: ditto

**Review by khancode — APPROVED** [review]
LGTM 👍 Added small nit comments that you can do before merging.

File: src/models/transactions/offerCancel.ts
Comment (JST5000) [reply src/models/transactions/offerCancel.ts]: For `tx.OfferSequence` since it's not optional I think the condition is actually:
`(typeof tx.OfferSequence !== 'number')`

Because OfferSequence isn't optional, we want this to throw an error when OfferSequence is undefined (And since typeof undefined !== number, that's the only condition we need)

**Review by JST5000 — APPROVED** [review]

File: src/models/transactions/transaction.ts
Comment (mvadari) [inline src/models/transactions/transaction.ts:7]: be uniform about semicolon use

**Review by mvadari — APPROVED** [review]

File: src/models/transactions/transaction.ts
Comment (mukulljangid) [reply src/models/transactions/transaction.ts]: Done

**Pull request merged**
